src/proper_target.erl

```diff
@@ -265,20 +303,35 @@ init([{Strategy, Data}]) ->
 handle_call(gen, _From, State) ->
   #state{strategy = Strat, target = Target, data = Data} = State,
   {NextValue, NewTarget, NewData} = Strat:next(Target, Data),
-  {reply, NextValue, State#state{target = NewTarget, data = NewData}};
+  Ret = case State#state.stateful of
+          true -> get_stateful_cmds(NextValue);
+          false -> NextValue
+        end,
+  {reply, Ret, State#state{target = NewTarget, data = NewData}};
+
 handle_call(shrinker, _From, State) ->
   #state{strategy = Strat, target = Target, data = Data} = State,
   Shrinker = Strat:get_shrinker(Target, Data),
   {reply, Shrinker, State};
+
 handle_call({init_target, Type}, _From, State) ->
   #state{strategy = Strat} = State,
   NextFun = proper_gen_next:from_proper_generator(Type),
   NewTarget = Strat:init_target(Type, NextFun),
   {reply, ok, State#state{target = NewTarget}};
+
+handle_call(init_stateful, _From, State) ->
+  {reply, ok, State#state{stateful = true}};
+
+handle_call({pdict, KVs}, _From, State) ->
+  lists:foreach(fun ({K, V}) -> put(K, V) end, KVs),
```

Comment:
List constructions are faster and more expressive here:
[put(K, V) || {K, V} <- KVs]

Also, we are using a gen_server an you still write stuff to the process dictionary. This should really be stored in the gen_server state instead. Please change

Reply:
> List constructions are faster and more expressive here:
> [put(K, V) || {K, V} <- KVs]

This is not a good argument.  IMO, the `lists:foreach/2` is the PropEr call here.
(In general, when you want to execute something for its side-effects only and not construct a _resulting_ list.  It makes the intention clear; a comprehension does not.)

Reply:
These values are needed, so that when someone is using `?SIZED` in a `?USERNF` neighborhood function, then the `size` is correctly provided (as the `$size` variable exists and gets modified in the main `proper`'s process dictionary, see [`proper:get_size()`](https://github.com/proper-testing/proper/blob/master/src/proper.erl#L660)).

Reply:
I'm mostly arguing against the usage of the process dictionary. It's local to the gen_server process and as such only accessible from within it. They should really be stored in the gen_server state. 

How do you read these values?

Reply:
They are being used during the call `proper_gen:safe_generate`, which then calls `proper:get_size`. So this is something unrelated to the targeted implementation, but is used by other parts of PropEr and I wanted a way to make sure that `?SIZED` generators worked as expexted when using a user provided NF. Without these updates, the gen_server does not obtain the updated $size, so the generation never grows in size.

Reply:
I think there is a misunderstanding here:
- I don't argue to have or not have the values stored
- the use of the process dictionary (`get/1` and `put/2`) does not make sense as every function that can access the dictionary also gets the `gen_server` state as an argument. So store these values in the state instead.
- If you read these values from outside the `gen_server` process (outside handle_... of this module) you are reading different values.

Reply:
Again, PropEr uses a `$size` variable in its process dictionary which is used whenever a `?SIZED` generator is generating values. Adding `size` to the `gen_server`'s state would not fix the problem, as while the call to `proper_gen:safe_generate` occurs within a `handle_call`, there is no way to let that function know of a new `$size` value produced by `proper:grow_size` calls.

To elaborate a bit further, if we exclude these updates to the `proper_target` `gen_server`'s process dictionary, then that has a value of `$size` equal to the `start_size`, which never gets updated. So if a user provided a neighborhood function with `?SIZED` in it, then the size would never grow.

Again, as mentioned below, all of this could be avoided by refactoring the main `proper` module to a `gen_server`, but this should be part of another PR.

It's not clear to me whether there is a better solution to the above problem for the moment, but I am open to suggestions.

Reply:
Wait, so the targeted generators are evaluated in the gen_server process and they are accessing the process dictonary for sized because that's how it was done in the main proper prcoess? :-(

Reply:
Exactly, that is why this was needed :/